lib/AST/ModuleNameLookup.cpp

```diff
@@ -254,194 +114,114 @@ class LookupVisibleDecls : public ModuleNameLookup<LookupVisibleDecls> {
+        == decls.end())
+      canReturnEarly = false;
+  }
 
-template <typename LookupStrategy>
-template <bool CRTPWorkaround>
-void ModuleNameLookup<LookupStrategy>::collectLookupResultsFromImports(
-    SmallVectorImpl<ValueDecl *> &decls,
-    ArrayRef<ModuleDecl::ImportedModule> reexports,
-    ModuleDecl::AccessPathTy accessPath,
-    const DeclContext *moduleScopeContext,
-    OverloadSetTy<CRTPWorkaround> &overloads) {
-
-  static_assert(
-      std::is_same<decltype(overloads),
-                   typename LookupStrategy::OverloadSetTy &>::value,
-      "Template params should be inferred.");
-
-  // Prefer scoped imports (those importing a specific name from a module, like
-  // `import func Swift.max`) to whole-module imports.
-  SmallVector<ValueDecl *, 8> unscopedValues;
-  SmallVector<ValueDecl *, 8> scopedValues;
-  for (auto next : reexports) {
-    // Filter any whole-module imports, and skip specific-decl imports if the
-    // import path doesn't match exactly.
-    ModuleDecl::AccessPathTy combinedAccessPath;
-    if (accessPath.empty()) {
-      combinedAccessPath = next.first;
-    } else if (!next.first.empty() &&
-               !ModuleDecl::isSameAccessPath(next.first, accessPath)) {
-      // If we ever allow importing non-top-level decls, it's possible the
-      // rule above isn't what we want.
-      assert(next.first.size() == 1 && "import of non-top-level decl");
-      continue;
-    } else {
-      combinedAccessPath = accessPath;
+  // If needed, search for decls in re-exported modules as well.
+  if (!canReturnEarly) {
+    auto &imports = ctx.getImportCache().getImportSet(moduleOrFile);
+
+    auto visitImport = [&](ModuleDecl::ImportedModule import,
+                           const DeclContext *moduleScopeContext) {
+      if (import.first.empty())
+        import.first = accessPath;
+      else if (!accessPath.empty() &&
+               !ModuleDecl::isSameAccessPath(import.first, accessPath))
+        return;
+
+      getDerived()->doLocalLookup(import.second, import.first, decls);
+      updateNewDecls(moduleScopeContext);
+    };
+
+    // If the ClangImporter's special header import module appears in the
+    // import set, we must visit it first.
+    ModuleDecl *headerImportModule = nullptr;
+    if (imports.hasHeaderImportModule()) {
+      if (auto *loader = ctx.getClangModuleLoader()) {
+        headerImportModule = loader->getImportedHeaderModule();
+        if (headerImportModule) {
+          ModuleDecl::ImportedModule import({}, headerImportModule);
+          visitImport(import, nullptr);
+        }
+      }
     }
 
-    auto &resultSet = next.first.empty() ? unscopedValues : scopedValues;
-    lookupInModule(resultSet, next.second, combinedAccessPath,
-                   moduleScopeContext);
-  }
+    for (auto import : imports.getTopLevelImports()) {
+      // A module appears in its own top-level import list; since we checked
+      // it already, skip it.
```

Comment:
It shouldn't.

Reply:
To be clear that comment refers to an artifact of how ImportSet is built. It's not talking about what getImportedModules() returns.

When building an ImportSet we compute the transitive closure of a set of ImportedModules. If you ask for the ImportSet of a source file, the generating set contains the module itself and the source file's imports. If you ask for the ImportSet of a module, the generating set is just the module itself.

Reply:
Ahh, okay. I guess that makes sense.